### ISSUE TITLE
Envoy listener doesn't support regex > 100 chars

### ISSUE THREAD

**Issue by tjhiggins** [hashicorp/consul#9809]
#### Overview of the Issue
Consul uses the envoy/go-control-plane v0.9.5 to perform the regex for the envoy listener response.
https://github.com/hashicorp/consul/blob/3b6ffc447b4d277afc26a85797fc73c3130333ad/agent/xds/response.go#L62

Envoy had a bug where the regex was capped at 100 chars. It appears the fix was merged into envoy/go-control-plane >=v0.9.7.
https://github.com/envoyproxy/envoy/pull/11187

This prevents the service from being able to communicate since the listener cannot be added.
The only workaround is to cap the service name/namespace to be under 100 chars. 

#### Reproduction Steps

Steps to reproduce this issue, eg:

1. Create a k8s cluster with 1 client nodes and 1 server nodes
1. Register a service with a long name 63 chars in the default namespace
1. Add wildcard intention to deny all traffic by default
1. Add intention allowing service to communicate with another service
1. View logs of envoy sidecar

### Log Fragments

Envoy sidecar logs
```console
[source/common/config/grpc_subscription_impl.cc:107] gRPC config for type.googleapis.com/envoy.api.v2.Listener rejected: Error adding/updating listener(s) public_listener:10.1.4.21:20000: 
regex '^spiffe://[^/]+/ns/default/dc/[^/]+/svc/examples--react-app--app--latest--component-cce5ff--lqkqseme$' 
RE2 program size of 102 > max program size of 100 set for the error level threshold. Increase configured max program size if necessary.
```

### Proposal
Upgrade envoy/go-control-plane >= v0.9.7

**Comment by jsanant** [comment]
@jsosulska - We are facing the same issue when we upgraded to consul v1.9.4. Can you please mark this as a bug? As this impacting our already existing service mesh.

**Comment by rboyer** [comment]
Blocked on https://github.com/hashicorp/consul/issues/10471

**Comment by danieleva** [comment]
We are running a patched consul binary to workaround the issue, adding a custom high value for maxRegexProgramSize so the wrong default `100` is not used: 
```
 		EngineType: &envoymatcher.RegexMatcher_GoogleRe2{
-			GoogleRe2: &envoymatcher.RegexMatcher_GoogleRE2{},
+			GoogleRe2: &envoymatcher.RegexMatcher_GoogleRE2{
+				MaxProgramSize: &wrappers.UInt32Value{Value: uint32(maxRegexProgramSize)},
+			},
 		},
```
I agree updating envoy/go-control-plane is the best long term solution but #10471 could take a long time to get fixed. 
Would you be interested in a patch similar to the snippet above?

**Comment by jkirschner-hashicorp** [comment]
Hi @danieleva,

We're glad you have a workaround for now and shared it for others to use if needed.

We considered your suggestion to accept a patch similar to the above. However, because there are several related issues, we decided it would be best to resolve holistically in #10741 rather than add/maintain several patches.

For anyone tracking this, consider subscribing to #10741 to stay apprised of any updates there.

**Comment by blake** [comment]
Hi @tjhiggins and @danieleva, Consul 1.12.0 now uses go-control-plane version 0.9.9.

Are you able to test with this latest Consul release and let us know whether you're still experiencing this issue?

Thanks.

**Comment by kmott** [comment]
@blake I updated to Consul v1.12.0, but I am still running into the problem--do I need to tweak a Consul config to allow an RE2 size > 100?

```
[2022-05-20 23:16:13.580][1][warning][config] [source/common/config/grpc_subscription_impl.cc:126] gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) public_listener:0.0.0.0:22485: regex '^spiffe://[^/]+/ns/default/dc/[^/]+/svc/ab-cde-01fwev12657q80kmc49pekfa2n-01g3ht1kg9pjrgx3n4kc5wtbjr$' RE2 program size of 102 > max program size of 100 set for the error level threshold. Increase configured max program size if necessary.
```

**Comment by danieleva** [comment]
Sorry for the late reply. Tested with v1.12.2 and envoy v1.22.2, unfortunately the issue is still there. After some digging I found https://github.com/envoyproxy/envoy/issues/18633 and https://github.com/envoyproxy/envoy/pull/2184 that clarify what's going on there. 
https://github.com/envoyproxy/envoy/pull/11187 was only the beginning, the config is deprecated but the value still checked against runtime config, and defaults to 100.
Adding a snippet like this to bootstrapTemplate allows you to set the value: 
```json
"layered_runtime": {
    "layers": [
      {
        "name": "static_layer_0",
        "static_layer": {
          "re2": {
            "max_program_size": {
              "error_level": 10000
            }
          }
        }
      }
    ]
  }
``` 
That can be achieved by overriding the entire template via proxy-defaults config entry, but it's cumbersome. I ended up  changing my patch to update `bootstrap_tpl.go`. Since the original assumption (upgrading go-control-plane) turned out to be ineffective to fix the bug, would you reconsider accepting a PR to add the runtime config directly in the default bootstrap template?

**Comment by blake** [comment]
Hi @danieleva, I believe PR #13629 incorporates a change to resolve this issue. The higher regex limit should at least be available in Consul 1.13.0.

**Comment by danieleva** [comment]
I didn't spot that PR. Yes that includes the change in bootstrapTemplate to increase the regex size. I tested it locally and it works as expected.  I'll keep my patch in our local 1.12 build, looking forward to 1.13 to be released :D.
I think this ticket can be closed now that #13629 is merged

**Comment by kmott** [comment]
@danieleva or @blake (my apologies for the @'s, I'm grasping at straws), I updated to Consul v1.17.0, and for some reason I hit this issue again.  Note that my deployment is using Nomad + Consul + Consul Connect, and it only seems to be one specific job that runs into the limit (have 100's of others that run fine and have a similar naming scheme).  I am happy to add the `proxy-defaults` override, but had a hard time deciphering where/how to do that in the docs--could you give me a little more info on how to do that for Nomad jobs?

Also, I am happy to open a new issue, if needed as well.  Thank you for your time.

```
Nov 29 16:11:15 nomad-prd-02 hab[806]: consul.nomad-prd(O): 2023-11-29T16:11:15.803-0800 [ERROR] agent.envoy.xds: got error response from envoy proxy: xdsVersion=v3 nonce=00000004
Nov 29 16:11:15 nomad-prd-02 hab[806]: consul.nomad-prd(O):   error=
Nov 29 16:11:15 nomad-prd-02 hab[806]: consul.nomad-prd(O):   | rpc error: code = Internal desc = Error adding/updating listener(s) public_listener:0.0.0.0:20289: regex '^spiffe://3a7b379c-67da-f15d-361c-0b2f714710b1.consul/ns/default/dc/[^/]+/svc/wpcc-01gbjf7fbek6612x6v2gjykmks-01ha7kwtjp6csvm6m918rdv9rj$' RE2 program size of 105 > max program size of 100 set for the error level threshold. Increase configured max program size if necessary.
```
